### PR TITLE
8237469: Inherited styles don't update when node is moved

### PR DESCRIPTION
**Pull request by DeanWookey**
Everything passes with the fix and 5 of the new tests fail without the fix.

removingThenAddingNodeToDifferentBranchGetsNewFontStyleTest
movingBranchToDifferentBranchGetsNewCssVariableTest
removingThenAddingNodeToDifferentBranchGetsCorrectInheritedValue
removingThenAddingNodeToDifferentBranchGetsIneritableStyle
movingNodeToDifferentBranchGetsNewFontStyleTest

I doubt this will cause a significant performance decrease. I think it's worth it for correctness. The only other thing is that I kept the fix to the minimum, but technically canReuseHelper now has a side effect. I could try rewrite some things if necessary, or maybe rename the method? Else leave it as is?

Originally introduced here: https://bugs.openjdk.java.net/browse/JDK-8090462/ https://github.com/openjdk/jfx/commit/834b0d05e7a2a8351403ec4a121eab312d80fd24#diff-9ec098280fa1aeed53c70243347e76ab

## Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

## Issue
[JDK-8237469](https://bugs.openjdk.java.net/browse/JDK-8237469): Inherited styles don't update when node is moved

## Approvers
 * David Grieve ([dgrieve](@dsgrieve) - **Reviewer**) **Note!** Review applies to 782f55e3375651e79b6a2ebe50dbbb5a86a69db0
 * Ajit Ghaisas ([aghaisas](@aghaisas) - **Reviewer**)
 * Kevin Rushforth ([kcr](@kevinrushforth) - **Reviewer**)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
I mentioned this in the JBS bug as well: I suspect that [JDK-8234877](https://bugs.openjdk.java.net/browse/JDK-8234877) has the same root cause as this one. Can you run the HelloCSS test program as described in that bug with your patch from this PR and see if it also fixes that bug?

**Comment by kevinrushforth** [comment]
This will need (at least) two reviewers. I request @aghaisas and @dsgrieve to review (I'll review it as well next week, once I'm done with the more urgent reviews).

**Comment by DeanWookey** [comment]
> I mentioned this in the JBS bug as well: I suspect that [JDK-8234877](https://bugs.openjdk.java.net/browse/JDK-8234877) has the same root cause as this one. Can you run the HelloCSS test program as described in that bug with your patch from this PR and see if it also fixes that bug?

I ran it and it seems to work. The button is always red in the red pane and always yellow in the yellow one.

File: modules/javafx.graphics/src/main/java/javafx/scene/CssStyleHelper.java [hunk 1]
```diff
@@ -131,7 +131,7 @@ static CssStyleHelper createStyleHelper(final Node node) {
             }
             node.styleHelper.cacheContainer.forceSlowpath = true;
             node.styleHelper.triggerStates.addAll(triggerStates[0]);
-            node.styleHelper.firstStyleableAncestor = findFirstStyleableAncestor(node);
```
Comment (dsgrieve) [inline modules/javafx.graphics/src/main/java/javafx/scene/CssStyleHelper.java]: In canReuseStyleHelper, there is this check
```
       // If the style maps are the same instance, we can re-use the current styleHelper if the cacheContainer is null.
        // Under this condition, there are no styles for this node _and_ no styles inherit.
        if (node.styleHelper.cacheContainer == null) {
            return true;
        }
```
And later in the same function is a check for (parent == null) that returns true. In both cases, firstStyleableAncestor will still point to the old first-styleable ancestor.

Comment (DeanWookey) [reply modules/javafx.graphics/src/main/java/javafx/scene/CssStyleHelper.java]: Thanks. For some reason I thought all the early exits out of the method returned false. I'll push a fix.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java [hunk 2]
```diff
@@ -0,0 +1,602 @@
+/*
+ * Copyright (c) 2020, 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (aghaisas) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java]: Minor : As this is a new file being added, there should be just one 2020 (The year in which file has been introduced)

**Comment by kevinrushforth** [comment]
I updated the PR title to match the JBS title, so that the commit message will be correct when this PR is integrated.

**Pull request closed without merging**